### ISSUE TITLE
New semantic analyzer: incompatible definition involving decorator not detected

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#6610]
In some cases of a function conditional definition (at least if the latter variant is a decorated function), there can be a false negative with the new semantic analyzer where we don't detect that the signatures are incompatible:

```py
from typing import Callable

def dec(f) -> Callable[[int], None]: pass

x = int()
if x:
    def f(x: str) -> None: pass
else:
    # No complaint about incompatible redefinition!
    @dec
    def f(): pass
```

**Comment by ilevkivskyi** [comment]
Now I get a correct error:
```
tstrdf.py:11: error: All conditional function variants must have identical signatures  [misc]
tstrdf.py:11: note: Original:
tstrdf.py:11: note:     def f(x: str) -> None
tstrdf.py:11: note: Redefinition:
tstrdf.py:11: note:     def f(int, /) -> None
```
